simpervisor/process.py

```diff
@@ -126,7 +126,12 @@ async def _restart_process_if_needed(self):
         This is a long running task that keeps running until the process
         exits. If we restart the process, `start()` sets this up again.
         """
-        retcode = await self.proc.wait()
+        # Since self.proc is a subprocess.Popen object, wait() method is a blocking
+        # call. To prevent the entire program being stuck at this point, we use
+        # the poll() method to check whether the child process is alive or not.
+        while self.proc.poll() is None:
+            await asyncio.sleep(0.1)
```

Comment:
I worry this would have a *huge* performance impact on running notebook server processes, as we would be adding sleep calls every 100ms per proxied process for the duration of the server running. Do you think instead we can still use asyncio's subprocess when on linux and fall back to regular one on Windows?

Reply:
Hi @yuvipanda, Thanks for your feedback.

I picked this approach to spin waiting from similar code in the `jupyter-client` project. Click [here ](https://github.com/jupyter/jupyter_client/blob/main/jupyter_client/provisioning/local_provisioner.py#L54-L55)to view their usage of it.
Yes, arguably there are going to be additional sleep events added to the event loop, however these do not manifest with any measurable slow down. Given that this is being used in the `jupyter-client` repository as well, it gives me some measure of confidence wrt to its performance.

Do you have a scenario in mind that I could use to test any performance fallout from the inclusion of this code?

I did consider branching the implementation based on platform, but that approach gets messy very quickly given the number of functions and code paths that need to branch.
Once again, I've drawn my inspiration of subprocess management with `subprocess.Popen` from the `jupyter-client` sources.

Reply:
Ah, seeing it in jupyter-client is very interesting and makes me feel much better about this approach :)

But, for the branching, what do you think of this approach where we have a wrapper class that just does two things - launches processes and waits for them. Then we can have two implementations of this class - one that uses asyncio directly, and one that uses popen + poll. This allows them to be unit tested differently as well, and we can instantiate the right class based on the platform. This keeps the code in simpervisor core clean, and hopefully leads to *fewer* branches than we have now. Would you be willing to give this a go?

Reply:
That sounds like a reasonable suggestion. However, I fear that introducing this approach would split the mental model for this repo, which might in turn lead to additional maintenance overheads. 

Perhaps we could try to implement this approach if use cases which require the ASYNCIO arise?

Reply:
@rashedmyt I'm wary of us having to do os-specific subprocess things throughout the codebase over time, and would very much rather instead have that isolated clearly in a specific place. Given that this package is very focused on spawning processes and checking if they are still alive, I think it's ok to introduce that layer here.

Reply:
hi @yuvipanda do you have an example of a repository which currently implements the design you envision? In the spirit of expedience, having an example would help us meet the design goals quickly. From what we understand of the codebase, developing a wrapper class with no branching code might be a somewhat large project to complete, review and deliver.

Reply:
@yuvipanda , thank you for the detailed response. We will take a pass at this and update the PR accordingly! 
We'll target having it for your review before the end of the second week of May.

Reply:
@prabhakk-mw glad to help!

Reply:
Hi @yuvipanda, I have updated the implementation to use the suggested style. Marking this thread as resolved. If there is any new feedback please open a new thread.